crates/containerd-shim-wasm/src/sandbox/containerd/client.rs

```diff
@@ -30,6 +32,9 @@ use crate::sandbox::oci::{self, WasmLayer};
 use crate::with_lease;
 
 static PRECOMPILE_PREFIX: &str = "runwasi.io/precompiled";
+// 16MB is the maximum size of a chunk that can be streamed.
+// Conservatively set the max to 15MB.
+static MAX_WRITE_CHUNK_SIZE_BYTES: i64 = 1024 * 1024 * 15;
```

Comment:
The default setting in containerd is here (16MB): https://github.com/containerd/containerd/blob/e53663cca75a3dd9c688a65a04f79350d6bb1fbd/defaults/defaults.go#L22

Though this is configurable in containerd.toml.
So I guess this chunking needs to be smaller than that to make room for other message overhead.


---


For a long time I've been wanting to add a file reference or pipe or something to the content store API so that this chunking could be skipped.

Reply:
ahhh if we change the DefaultMaxRecvMsgSize in containerd.toml to be something smaller (4mb), then the hard coded 15MB chunk size would fail in runwasi, right?

Reply:
Confirmed that this is the case, updating the containerd config would cause this to fail:
```sh
ctr: failed to copy: failed to send write: grpc: received message larger than max (1048582 vs. 200000): unknown
```

Reply:
Maybe our client can handle this error and reduce the chunking until it finds something that works?

Reply:
Raised an issue #509